### PR TITLE
Health checks

### PR DESCRIPTION
**Pull request by NathanielRose**
@erikschlegel @jcjimenez 
Working and validated .V1 of Health Checks.
Please review local-deploy.sh & docs/local-deploy.md
[Issue 166](https://github.com/CatalystCode/project-fortis-pipeline/issues/166)

### CODE REVIEW

File: docs/local-deploy.md
Comment (Smarker) [inline docs/local-deploy.md]: Could say `your machine is correctly setup...`

File: docs/local-deploy.md
Comment (Smarker) [inline docs/local-deploy.md]: You don't need the ending ##

File: docs/local-deploy.md
Comment (Smarker) [inline docs/local-deploy.md]: `set up variables for development environment`

File: docs/local-deploy.md
Comment (Smarker) [inline docs/local-deploy.md]: You probably don't need to say `Run the following command in the terminal.`

File: local-deploy.sh
Comment (Smarker) [inline local-deploy.sh]: `..following variables`

spelling of `currently`

File: local-deploy.sh
Comment (Smarker) [inline local-deploy.sh]: `please validate that you have ...`

File: local-deploy.sh
Comment (Smarker) [inline local-deploy.sh]: probably don't need these separator comments

File: local-deploy.sh
Comment (Smarker) [inline local-deploy.sh]: did you fix this? if so, I'd remove the comment

**Review by Smarker — COMMENTED** [review]
Someone else should take a look at the shell script, just to make sure it makes sense and is correct. I mainly looked at how readable your docs were and what comments you used.

**Review by jcjimenez — APPROVED** [review]
LGTM  - wondering if we should convert the script into a Dockerfile.

File: docs/local-deploy.md
Comment (NathanielRose) [reply docs/local-deploy.md]: This is from the readme from prior directly. I made the edits though.

File: local-deploy.sh
Comment (c-w) [inline local-deploy.sh]: Nit: might as well quote the echoed string to make this look more expected and to be consistent in all echo call sites: `echo "found JAVA executable in PATH"`

File: local-deploy.sh
Comment (c-w) [inline local-deploy.sh]: Nit: bash has built-in support for arithmetic, no need to go to awk for this:

```sh
# bashism
if (( vCompare_cassandra > minCassandra )); then

# posix
if [ "$vCompare_cassandra" -gt "$minCassandra" ]; then
```

File: local-deploy.sh
Comment (c-w) [inline local-deploy.sh]: Nit: bash has built-in support for string slicing, no need to go to awk for this:

```sh
vCompare_cassandra="${vCompare_cassandra:0:3}"
```

File: local-deploy.sh
Comment (c-w) [inline local-deploy.sh]: FYI: if you use `command -v spark-shell` it'll be posix-compliant and not relying on a bashism.

One of the common helper functions I have in my scripts for this use-case is:

```sh
command_exists() { command -v "$1" >/dev/null 2>&1; }
```

File: local-deploy.sh
Comment (c-w) [reply local-deploy.sh]: Same elsewhere in this file.

File: local-deploy.sh
Comment (c-w) [inline local-deploy.sh]: FYI: for simple string deletions like this, you can use tr which is more explicit since its capabilities are more restricted:

```sh
tr -d 'A-Za-z ._-'
```

File: local-deploy.sh
Comment (c-w) [inline local-deploy.sh]: FYI: for simple column selects, you can use cut which is more explicit than full-blown awk since its capabilities are more restricted:

```
cut -d' ' -f5
```

File: local-deploy.sh
Comment (c-w) [inline local-deploy.sh]: This version number parsing is happening in a few places; would it be worth extracting this to a method?

```sh
to_major_version() { echo "$1" | sed -e 's/[A-Za-z ._-]//g'| awk '{print substr($0,0,3)}'); }
```

File: local-deploy.sh
Comment (c-w) [inline local-deploy.sh:1]: Now that we have more shell scripts in this repository, could you please add [shellcheck](https://www.shellcheck.net) to our Travis CI script? See [here](https://github.com/CatalystCode/project-fortis-pipeline/blob/3def67ae2168cb687bcbbe2dbaf6a0821e4b8403/.travis.yml#L4-L12) for a sample that you can copy.

By the way: I love that your script is shellcheck-clean already. Great job; definitely demonstrates above-average bash skills :)

**Review by c-w — APPROVED** [review]
Good change from a code point of view. A few pointers to maybe make the bash code use some more idiomatic/posix features.

A more general question though: if it's such a pain to set up this project, would be benefit from just providing a dockerfile that developers can run to have an environment with all the dependencies set up for good?

**Comment by NathanielRose** [comment]
@erikschlegel @kevinhartman Looking to get final approvals to close pr

File: docs/local-deploy.md
Comment (kevinhartman) [inline docs/local-deploy.md]: Should the stuff above this line be here? It looks like a copy from the README.md in the root of this repo?

File: docs/local-deploy.md
Comment (kevinhartman) [inline docs/local-deploy.md]: I'd move this to `docs/developer/local-deploy.md` since the developer-centric docs have recently been moved there.

File: docs/local-deploy.md
Comment (kevinhartman) [inline docs/local-deploy.md]: You can refer the reader to the [environment documentation](https://github.com/CatalystCode/project-fortis-pipeline/blob/master/docs/developer/environment.md) here with a hyperlink. Once you move this file to the `developer` folder, you can link it like this: `See [Environment documentation](environment.md) for more details.`

File: docs/local-deploy.md
Comment (kevinhartman) [inline docs/local-deploy.md]: It looks like a lot of these are for a specific Linux distro. Since Fortis development is cross-platform (Mac, Windows, Linux), it might be useful to note exactly which OS these instruction are for, and that the user may need to do things differently for any other environment. **Most importantly, we should note the versions of these dependencies, perhaps in a table.**

File: local-deploy.sh
Comment (kevinhartman) [inline local-deploy.sh:172]: Do we actually have a hard dependency on Cassandra? As a developer, I don't need to run a Cassandra instance on my local machine. I can point to one that's elsewhere. I'd recommend removing this, in order to defer the that choice to the developer. Perhaps a warning would be appropriate?

**Review by kevinhartman — APPROVED** [review]
LGTM after comments addressed.

General question, is this check script to be run on developer machines in order to verify that their environment is set up for dev work, or is it also to be used in production?

In the developer-only case, some tools are not needed- for example, `kubectl` is only used if you need to connect to a remote cluster. If you're just running locally, you won't have a dependency on it.

**Pull request merged**
